core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorSplitManager.java

```diff
@@ -47,6 +61,21 @@ default ConnectorSplitSource getSplits(
         return getSplits(transaction, session, table, splitSchedulingStrategy, dynamicFilter);
     }
 
+    default ConnectorSplitSource getSplits(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            SplitSchedulingStrategy splitSchedulingStrategy,
+            DynamicFilter dynamicFilter,
+            Constraint constraint,
+            RetryMode retryMode)
```

Comment:
>BTW For cases where splits are not idempotent, do we have an idea how this could work?

I do not think we did think much about that already. First idea that comes to my mind is to make splits usage idempotent. But I don't know how easy that would be.

Reply:
> Also, for query-level retries, do we need to bother ConnectorSplitManager at all?

this is a good point. For query level retries I think we are good. As we restart query from scratch and enumerate splits once again. I will change logic here to only throw if retryMode is TASKS and clarify in Javadoc.

Reply:
> I will change logic here to only throw if retryMode is TASKS and clarify in Javadoc.

then we shouldn't pass `RetryMode` at all, and use a boolean instead. or another enum.

> First idea that comes to my mind is to make splits usage idempotent. But I don't know how easy that would be.

Think of a connector which does some prep work in SplitManager and then creates splits that consume that work's results. In such design, i don't see how idempotency could be "added". 🤷‍♂️  for now.
Maybe a connector  like that simply requires re-architecting to work with retries.

Reply:
After discussion I introduced another 2-value enum here:

```
public enum SplitUseMode
{
    SINGLE_USE,
    REPEATED_CONCURRENT_USE,
}
```